TDM hardware connection (#476)
XanaduAI/strawberryfields: .github/CHANGELOG.md

<h3>New features since last release</h3>

> * `TDMProgram`s can now be compiled and submitted via the API.
>   [(#476)](https://github.com/XanaduAI/strawberryfields/pull/476)

Reviewer: ```suggestion
* `TDMProgram` objects can now be compiled and submitted via the API.
  [(#476)](https://github.com/XanaduAI/strawberryfields/pull/476)
```
(ReST doesn't like characters occurring after the backticks)

<h3>Improvements</h3>

> * `copies` are renamed `shots` inside of `TDMProgram` and is now passed via
>   the `eng.run()` method instead of via the program context.
>   [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

Reviewer: ```suggestion
* The `copies` option when constructing a `TDMProgram` have been removed. Instead, the number of
  copies of a TDM algorithm can now be set by passing the `shots` keyword argument to 
  the `eng.run()` method.
  [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)
```

>   ```pycon
>     >>> with prog.context([1, 2], [3, 4]) as (p, q):
>     ...     ops.Sgate(0.7, 0) | q[1]
>     ...     ops.BSgate(p[0]) | (q[0], q[1])
>     ...     ops.MeasureHomodyne(p[1]) | q[0]
> 
>     >>> eng = sf.Engine("gaussian")
>     >>> results = eng.run(prog, shots=3)
>   ```

Reviewer: ```suggestion
  ```pycon
  >>> with prog.context([1, 2], [3, 4]) as (p, q):
  ...     ops.Sgate(0.7, 0) | q[1]
  ...     ops.BSgate(p[0]) | (q[0], q[1])
  ...     ops.MeasureHomodyne(p[1]) | q[0]
  >>> eng = sf.Engine("gaussian")
  >>> results = eng.run(prog, shots=3)
  ```
```

>   Furthermore, `TDMProgram.unrolled_circuit` only contains the single-shot
>   unrolled circuit, while the unrolling including shots is returned by the
>   `TDMProgram.unroll()` method, and is strictly contained to this method
>   (and thus not an attribute of the `TDMProgram` class any longer).

Reviewer: ```suggestion
  Furthermore, the `TDMProgram.unrolled_circuit` attribute now only contains the single-shot
  unrolled circuit. Unrolling with multiple shots can still be specified via the `unroll` method:
  `TDMProgram.unroll(shots=60)`.
```

* The `reshape_samples` function is updated to return a dictionary where
  each key correspond to a spatial mode and the values have shape ``(shots,
  timebins)``.

>   [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

Reviewer: Since this is an internal implementation detail, perhaps it's not worth including?

> * The `Result.samples` is updated for TDM programs to return samples of
>   shape `(shots, spatial modes, timebins)` instead of `(shots, spatial
>   modes * timebins)`.
>   [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

Reviewer: ```suggestion
* The `Result.samples` returned by TDM programs has been updated to return
  samples of shape `(shots, spatial modes, timebins)` instead of `(shots, spatial
  modes * timebins)`.
  [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)
```

>   [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

Reviewer: This is another reason why I suggest not using a holding branch for future projects --- so many changelog entries refer to the same (large) PR 😆

> * A sample post-processing function is added that allows user's to move
>   vacuum mode measurements from the first shots to the last shots, and
>   potentially crop out the final shots containing these measurements.
>   [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

Reviewer: ```suggestion
* A sample post-processing function is added that allows users to move
  vacuum mode measurements from the first shots to the last shots, and
  potentially crop out the final shots containing these measurements.
  [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)
```

Potentially also not to needed to be included, since it is more of an internal implementation detail?

> * `TDMProgram`s can be converted into Blackbird programs and vice versa.
>   [(#476)](https://github.com/XanaduAI/strawberryfields/pull/476)

Reviewer: ```suggestion
* `TDMProgram` objects can now be serialized into Blackbird scripts, and vice versa.
  [(#476)](https://github.com/XanaduAI/strawberryfields/pull/476)
```

<h3>Breaking changes</h3>

<h3>Bug fixes</h3>

* Fixed issue with `reshape_samples` where the samples were sometimes
  reshaped in the wrong way.
  [(#489)](https://github.com/XanaduAI/strawberryfields/pull/489)

* The list of modes is now correctly added to the Blackbird program when using the
  `io.to_blackbird` function.
  [(#476)](https://github.com/XanaduAI/strawberryfields/pull/476)

<h3>Documentation</h3>

<h3>Contributors</h3>

This release contains contributions from (in alphabetical order):

> Theodor Isacsson, Fabian Laudenbach, Nicolas Quesada

Reviewer: ```suggestion
Theodor Isacsson, Josh Izaac, Fabian Laudenbach, Nicolas Quesada, Antal Száva.
```
